### DISCUSSION TITLE
video autoselect add addon and music video

### DISCUSSION THREAD

**Discussion by nico1080** [mattblovell/kodi_panel#26] [Ideas]
Kodi also have a section for  video addon and musicvideo.

Currently the autoselect feature will not detect these kodi section.
Video addon detection should work with `info["Player.Filenameandpath"].startswith("plugin://")`
musicvideo detection with `VideoPlayer.Artist` infolabel but I do not use it so I can’t check it.

I do not think there is mor kodi section for video

**Comment by mattblovell** [comment]
See kodi_panel v1.24.

The existing video selection if-elif heuristic was moved to `video_select_default()`, pointed to by the variable VIDEO_SELECT_FUNC.

A user's invoking script can override that variable like so:

   kodi_panel_display.VIDEO_SELECT_FUNC = my_selection_func

The function is passed the dictionary of InfoLabels retrieved from Kodi.  

(Note that Python provides a mechanism to permit for shorter namespace references.  Look for "import as" documentation.  Shortening the name for kodi_panel_display would permit for less verbose override assignments.)

I think the full customization that is now possible for the selection function addresses this idea.

**Comment by nico1080** [comment]
I am not sur to understand how this work.
Where is the `my_selection_func` defined? In a separate python script?
To start it, one would need a .sh script that call for kodi_panel_display script with variable that point to other variable script?

When I suggested this feature, I thought that the `setup.toml` would have a "Custom video autoselect heuristic script path"  that pointed out to a separate script defined by user

I also see you added autoselection for audio section. Does it benefit the same feature?

**Comment by mattblovell** [comment]
> Where is the my_selection_func defined? In a separate python script?

The override would in the script that contains the call to `kodi_panel_display.main()`, e.g.
- kodi_panel_fb.py or
- kodi_panel_ili9341.py or
- kodi_panel_ili9486.py

The above are the scripts that a user directly invokes, so that is where execution first starts.  

If you examine the current kodi_panel_fb.py, you will see this section of code prior to the `main()` invocation:

     # Setup PWM
     if config.settings["USE_HW_PWM"]:
         os.system("echo 0 > /sys/class/pwm/pwmchip0/export")
         sleep(0.150)
         freq_cmd   = "echo " + str(config.settings["HW_PWM_FREQ"]) + " > /sys/class/pwm/pwmchip0/pwm0/period"
         period_cmd = "echo " + str(int(config.settings["HW_PWM_FREQ"] *
                                        config.settings["HW_PWM_LEVEL"])) + " > /sys/class/pwm/pwmchip0/pwm0/duty_cycle"
         os.system(freq_cmd)
         os.system(period_cmd)
         screen_on_pwm()
         kodi_panel_display.screen_on  = screen_on_pwm
         kodi_panel_display.screen_off = screen_off_pwm


Those last two assignment statements serve as overrides for the default `screen_on()` and `screen_off()` functions, replacing them with two functions that are defined earlier in the kodi_panel_fb.py file.

A similar approach is what's intended for a heuristic function override.  I think the only "tricky part" is that references to the enumerated class may have to be **qualified** with the relevant namespace.  That is, an assignment to `new_mode` may have to look something like this:

    new_mode = kodi_panel_display.VDisplay["V_LIVETV"]

Other than that, since the dictionary of InfoLabels gets passed in, I think the if-elif tree would look similar to what's in the default function.

> I also see you added autoselection for audio section. Does it benefit the same feature?

It should behave identically, and also permit for user customization of the heuristic function.  The default selection function on the audio side is just empty (it returns the current `audio_dmode`).

**Comment by nico1080** [comment]
Ok, I think got it, I will try it this weekend.

I have to write in my `kodi_panel_fb.py`
`kodi_panel_display.VIDEO_SELECT_FUNC = custom_video_detection
`
and also define the custom_video_detection code in `kodi_panel_fb.py`
```
def custom_video_detection(info):
    if (info["Player.Filenameandpath"].startswith("pvr://recordings") and
        "V_PVR" in VIDEO_LAYOUT):
        new_mode = VDisplay["V_PVR"]     # PVR TV shows
    elif (info["Player.Filenameandpath"].startswith("pvr://channels") and
          "V_LIVETV" in VIDEO_LAYOUT):
        new_mode = VDisplay["V_LIVETV"]  # live TV
    elif (info["VideoPlayer.TVShowTitle"] != '' and
          "V_TV_SHOW" in VIDEO_LAYOUT):
        new_mode = VDisplay["V_TV_SHOW"] # library TV shows
    elif (info["VideoPlayer.OriginalTitle"] != '' and
          "V_MOVIE" in VIDEO_LAYOUT):
        new_mode = VDisplay["V_MOVIE"]   # movie
    else:
        # use the default mode specified from setup
        new_mode = VDisplay[config.settings["VLAYOUT_INITIAL"]]

    return new_mode
```
